public/app/features/plugins/datasource_srv.ts

```diff
@@ -51,13 +54,28 @@ export class DatasourceSrv implements DataSourceService {
       this.settingsMapById[dsSettings.id] = dsSettings;
     }
 
+    for (const ds of Object.values(this.runtimeDataSources)) {
+      this.datasources[ds.uid] = ds;
+      this.settingsMapByUid[ds.uid] = ds.instanceSettings;
+    }
+
     // Preload expressions
     this.datasources[ExpressionDatasourceRef.type] = expressionDatasource as any;
     this.datasources[ExpressionDatasourceUID] = expressionDatasource as any;
     this.settingsMapByUid[ExpressionDatasourceRef.uid] = expressionInstanceSettings;
     this.settingsMapByUid[ExpressionDatasourceUID] = expressionInstanceSettings;
   }
 
+  registerRuntimeDataSource(entry: RuntimeDataSourceRegistration): void {
+    if (this.runtimeDataSources[entry.dataSource.uid]) {
```

Comment:
Seems like this should not only be unique inside `runtimeDataSources` but also inside `datasources`. Or is the intentions to be able to overwrite existing data sources?

Reply:
yea, ideally it should be unique in both. think in scenes I had a way to ensure that runtime data source ids never conflicted with plugin data source plugin ids